rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+from __future__ import annotations
+import typing
+from abc import abstractmethod
+from typing import Optional, Text, Dict, List, Union, Iterable, Tuple, Any
+from collections.abc import ValuesView, KeysView
+import copy
+
+from rasa.engine.graph import GraphComponent
+
+if typing.TYPE_CHECKING:
+    from rasa.engine.storage.storage import ModelStorage
+    from rasa.engine.storage.resource import Resource
+    from rasa.engine.graph import ExecutionContext
+from rasa.shared.core.domain import Domain, SubState
+from rasa.shared.core.events import ActionExecuted, UserUttered, Event
+from rasa.shared.core.trackers import DialogueStateTracker
+from rasa.shared.core.training_data.structures import StoryGraph
+from rasa.shared.nlu.constants import ACTION_NAME, ACTION_TEXT, ENTITIES, INTENT, TEXT
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.nlu.training_data.features import Features
+
+# TODO: cacheable version of `CoreFeaturizationPrecomputations` and new
+# `collect_and_cache` function for `CoreFeaturizationCollector` to be used during
+# train (caveat: we're not able to dynamically switch on/off caching - we could only
+# change the recipe such that the non-caching prediction-time method would be used
+# to collect)
```

Comment:
Would be cool if we don't need to run any pre-computation again if we don't change the featurization pipeline components but only want to train/test different combinations of policies. So during policy training (and in the basic recipe), the output of the last component of the pipeline (called `*Collector` at the moment) could produce some cacheable output. Do you think this could work?

Reply:
ah yeah, sure 👍🏻 You can do by implementing the `Cacheable` protocol for the collector's output 👍🏻